pkg/storage/runtime.go

```diff
@@ -0,0 +1,418 @@
+	names, err = r.image.GetStore().GetNames(container.LayerID)
+	if err != nil {
+		return ContainerInfo{}, err
+	}
+	names = append(names, layerName)
+	err = r.image.GetStore().SetNames(container.LayerID, names)
+	if err != nil {
+		return ContainerInfo{}, err
+	}
+
+	// Find out where the container work directories are, so that we can return them.
+	containerDir, err := r.image.GetStore().GetContainerDirectory(container.ID)
+	if err != nil {
+		return ContainerInfo{}, err
+	}
+	if metadata.Pod {
+		logrus.Debugf("pod sandbox %q has work directory %q", container.ID, containerDir)
+	} else {
+		logrus.Debugf("container %q has work directory %q", container.ID, containerDir)
+	}
+
+	containerRunDir, err := r.image.GetStore().GetContainerRunDirectory(container.ID)
+	if err != nil {
+		return ContainerInfo{}, err
+	}
+	if metadata.Pod {
+		logrus.Debugf("pod sandbox %q has run directory %q", container.ID, containerRunDir)
+	} else {
+		logrus.Debugf("container %q has run directory %q", container.ID, containerRunDir)
+	}
+
+	return ContainerInfo{
+		ID:     container.ID,
+		Dir:    containerDir,
+		RunDir: containerRunDir,
+	}, nil
+}
+
+func (r *runtimeService) CreatePodSandbox(metadata RuntimeContainerMetadata) (ContainerInfo, error) {
+	metadata.Pod = true
+	return r.createContainerOrPodSandbox(metadata, metadata.PodID)
+}
+
+func (r *runtimeService) CreateContainer(metadata RuntimeContainerMetadata, containerID string) (ContainerInfo, error) {
+	metadata.Pod = false
+	return r.createContainerOrPodSandbox(metadata, containerID)
+}
+
+func (r *runtimeService) RemovePodSandbox(idOrName string) error {
+	container, err := r.image.GetStore().GetContainer(idOrName)
+	if err != nil {
+		if err == storage.ErrContainerUnknown {
+			return ErrInvalidSandboxID
+		}
+		return err
+	}
+	err = r.image.GetStore().DeleteContainer(container.ID)
+	if err != nil {
+		logrus.Debugf("failed to delete pod sandbox %q: %v", container.ID, err)
+		return err
+	}
+	return nil
+}
+
+func (r *runtimeService) DeleteContainer(idOrName string) error {
+	container, err := r.image.GetStore().GetContainer(idOrName)
+	if err != nil {
+		if err == storage.ErrContainerUnknown {
+			return ErrInvalidContainerID
+		}
+		return err
+	}
+	err = r.image.GetStore().DeleteContainer(container.ID)
+	if err != nil {
+		logrus.Debugf("failed to delete container %q: %v", container.ID, err)
+		return err
+	}
+	return nil
+}
+
+func (r *runtimeService) SetContainerMetadata(idOrName string, metadata RuntimeContainerMetadata) error {
+	mdata, err := json.Marshal(&metadata)
+	if err != nil {
+		logrus.Debugf("failed to encode metadata for %q: %v", idOrName, err)
+		return err
+	}
+	return r.image.GetStore().SetMetadata(idOrName, string(mdata))
+}
+
+func (r *runtimeService) GetContainerMetadata(idOrName string) (RuntimeContainerMetadata, error) {
+	metadata := RuntimeContainerMetadata{}
+	mdata, err := r.image.GetStore().GetMetadata(idOrName)
+	if err != nil {
+		return RuntimeContainerMetadata{}, err
```

Comment:
Sure; json.Unmarshal checks for well-formedness before parsing, so we shouldn't need to worry about returning partially-parsed results when there's a parsing error.